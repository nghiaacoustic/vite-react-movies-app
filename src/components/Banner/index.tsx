import React from 'react'
import { useParams, Link } from 'react-router-dom'
import { Movie } from '@/models'
import { BannerSkeleton } from '@/components'
import { formatDate, getImage, roundedNumber } from '@/utils/index.utils'
import { FaStar } from 'react-icons/fa'

export type BannerProps = {
  movie?: Movie
  isLoading: boolean
}

export const Banner: React.FC<BannerProps> = ({ movie, isLoading }) => {
  const { id } = useParams<{ id: string }>()

  if (isLoading) {
    return <BannerSkeleton />
  }

  return (
    <div
      className='relative mobile:h-[500px] tablet:h-[600px] desktop:h-[700px] monitor:h-[1080px] text-white-primary flex items-end justify-start p-8 !m-0'
      style={{
        backgroundImage: `url(${getImage(movie?.backdrop_path!, true)})`,
        backgroundSize: 'cover',
        backgroundPosition: 'center center'
      }}
      data-testid='banner'
    >
      <div className='bg-gradient-to-t from-black-primary via-transparent to-transparent absolute inset-0 z-0'></div>
      <div className='relative z-10 max-w-2xl space-y-4 tablet:text-xl monitor:text-3xl'>
        <h1 className='tablet:text-3xl text-2xl monitor:text-5xl font-bold mb-4'>
          {movie?.title}
        </h1>
        <p className='tablet:text-xl monitor:text-2xl text-sm mb-6'>
          {movie?.overview}
        </p>

        {!id ? (
          <div>
            <Link
              className='text-lg px-6 py-2 bg-red-600 hover:bg-red-700 text-white-primary rounded transition-colors ease-in-out duration-100'
              to={`/movie/${movie?.id}`}
            >
              More Info
            </Link>
          </div>
        ) : (
          <div className='space-y-2'>
            <p className='text-md'>
              <span className='font-semibold'>Release Date:</span>{' '}
              {formatDate(movie?.release_date!)}
            </p>
            <p className='text-md'>
              <span className='font-semibold'>Genres:</span>{' '}
              {movie?.genres?.map((genre) => genre.name).join(', ')}
            </p>
            <p className='text-md'>
              <span className='font-semibold'>Runtime:</span> {movie?.runtime}{' '}
              minutes
            </p>
            <div className='text-md flex-items-center gap-1'>
              <span className='font-semibold'>Rating:</span>
              <div className='flex-items-row gap-1'>
                <FaStar style={{ color: 'yellow' }} />
                <div className='text-lg font-bold text-yellow-400'>
                  {roundedNumber(movie?.vote_average ?? 0)}
                </div>
                ({movie?.vote_count} votes)
              </div>
            </div>
          </div>
        )}
      </div>
    </div>
  )
}

export default Banner
