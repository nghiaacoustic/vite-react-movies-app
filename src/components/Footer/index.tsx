export const Footer: React.FC = () => {
  return (
    <footer className='p-4 text-center bg-black-primary'>
      <div>
        <p>Â© FPT Software, Inc.</p>
      </div>
    </footer>
  )
}
