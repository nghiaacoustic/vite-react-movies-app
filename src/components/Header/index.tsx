import { useEffect, useState } from 'react'
import { NavLink } from 'react-router-dom'
import FPTLogo from '@/assets/fpt-logo.png'

const headerNav = [
  { path: '/', label: 'Home' },
  { path: '/movies/latest', label: 'Movies' }
]

export const Header: React.FC = () => {
  const [isScrolled, setIsScrolled] = useState(false)

  useEffect(() => {
    const handleScroll = () => {
      setIsScrolled(window.scrollY > 0)
    }
    window.addEventListener('scroll', handleScroll)

    return () => window.removeEventListener('scroll', handleScroll)
  }, [])

  return (
    <header
      className={`fixed flex-items-center justify-between py-4 px-8 w-full ${isScrolled ? 'bg-black-primary' : 'bg-gradient-to-r from-black-secondary via-transparent to-transparent'} transition duration-300 ease-in-out top-0 z-50`}
    >
      <div className='flex-items-center'>
        <NavLink to='/'>
          <img src={FPTLogo} alt='Logo' className='h-10 mr-10 cursor-pointer' />
        </NavLink>
        <nav className='ml-auto'>
          <ul className='flex space-x-4'>
            {headerNav.map((item) => (
              <li key={item.label}>
                <NavLink
                  to={item.path}
                  className={({ isActive }) =>
                    `${isActive ? 'font-bold text-white-primary' : 'font-thin text-white-secondary'} text-xl text-white-secondary hover:text-white-hover`
                  }
                >
                  {item.label}
                </NavLink>
              </li>
            ))}
          </ul>
        </nav>
      </div>
    </header>
  )
}
