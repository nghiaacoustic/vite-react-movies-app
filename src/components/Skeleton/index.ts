export * from './BannerSkeleton'
export * from './CardSkeleton'
export * from './ListCardSkeleton'
