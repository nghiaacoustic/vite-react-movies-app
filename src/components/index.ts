export * from './Header'
export * from './Footer'
export * from './Loader'
export * from './MovieCard'
export * from './MovieList'
export * from './ProfileCard'
export * from './Slider'
export * from './Banner'
export * from './Skeleton'
export * from './ScrollToTop'
