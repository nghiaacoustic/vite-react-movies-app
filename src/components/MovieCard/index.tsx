import React from 'react'
import { Link } from 'react-router-dom'
import { Movie } from '@/models'
import { FaStar } from 'react-icons/fa'
import { IoIosPlayCircle } from 'react-icons/io'
import { CardSkeleton } from '@/components'
import {
  capitalizeFirstLetter,
  formatDate,
  getImage,
  roundedNumber
} from '@/utils/index.utils'

export type MovieCardProps = {
  movie: Movie
  isLoading: boolean
}

export const MovieCard: React.FC<MovieCardProps> = ({ movie, isLoading }) => {
  if (isLoading && !movie) {
    return <CardSkeleton />
  }

  const { id, title, vote_average, release_date, poster_path } = movie
  return (
    <div
      data-testid='movie-card'
      className='max-w-sm min-w-[236px] w-full overflow-hidden shadow-md rounded-lg h-[350px] mx-auto p-2'
    >
      <Link to={`/movie/${id}`}>
        <div className='overflow-hidden rounded-xl relative flex-items-col transform hover:-translate-y-2 transition ease-in-out duration-500 shadow-lg hover:shadow-2xl movie-item text-white-primary h-[350px]'>
          <div className='absolute inset-0 z-10 transition duration-300 ease-in-out bg-gradient-to-t from-black-primary via-gray-900 to-transparent'></div>
          <div
            className='relative h-full cursor-pointer group z-10 px-10 pt-10 space-y-6 movie_info'
            data-lity=''
          >
            <div className='flex-items-col justify-between h-full'>
              <div className='h-40' />
              <div className='flex flex-col space-y-6 detail_info h-full'>
                <div className='flex flex-col space-y-2 inner w-52'>
                  <div
                    className='relative flex-items-center w-min flex-shrink-0 p-1 text-center text-white-primary bg-red-primary rounded-full group-hover:bg-red-hover transition-colors'
                    data-unsp-sanitized='clean'
                  >
                    <IoIosPlayCircle className='w-10 h-10' />
                    <div className='absolute transition opacity-0 duration-500 ease-in-out transform group-hover:opacity-100 group-hover:translate-x-16 text-xl font-bold text-white-primary group-hover:pr-2'>
                      Details
                    </div>
                  </div>
                  <h1 className='text-xl font-bold text-white-primary'>
                    {capitalizeFirstLetter(title)}
                  </h1>
                </div>

                <div className='flex flex-col justify-end flex-1 pb-4 '>
                  <div className='flex-items-row space-x-2 bg-gradient-to-t from-black-primary via-gray-900 to-transparent p-2 rounded-lg'>
                    <FaStar style={{ color: 'yellow' }} />
                    <div className='text-lg font-bold text-yellow-400'>
                      {roundedNumber(vote_average)}
                    </div>
                  </div>
                  <div className='flex-items-row space-x-2 mt-2'>
                    <div className='text-lg font-bold text-white-primary'>
                      {formatDate(release_date)}
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <img
            className='absolute inset-0 transform w-full -translate-y-4 object-cover h-full'
            src={getImage(poster_path!)}
            style={{ filter: 'grayscale(0)' }}
            alt='poster-image'
          />
        </div>
      </Link>
    </div>
  )
}
