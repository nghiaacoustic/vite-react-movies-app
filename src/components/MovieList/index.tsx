import React from 'react'
import { ListCardSkeleton, MovieCard } from '@/components'
import { Movie } from '@/models'
import { capitalizeFirstLetter } from '@/utils/index.utils'

type MovieListProps = {
  movies: Movie[]
  isLoading: boolean
  title: string
}

export const MovieList: React.FC<MovieListProps> = ({
  movies,
  isLoading,
  title
}) => {
  if (isLoading) {
    return <ListCardSkeleton count={20} />
  }

  return (
    <>
      <h1 className='font-bold text-4xl px-8 pt-10 pb-4 text-center'>
        {capitalizeFirstLetter(title)} Movies
      </h1>
      <div className='flex-full-center flex-row gap-2  mx-auto overflow-hidden flex-wrap'>
        {movies?.map((movie) => (
          <div key={movie.id}>
            <MovieCard movie={movie} isLoading={isLoading} />
          </div>
        ))}
      </div>
    </>
  )
}
