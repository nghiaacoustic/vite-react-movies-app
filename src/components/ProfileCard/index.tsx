import React from 'react'
import { getImage } from '@/utils/index.utils'
import { CardSkeleton } from '@/components'
import { CreditProfileInterface } from '@/models'

type ProfileCardProps = {
  profile: CreditProfileInterface
  isLoading: boolean
}

export const ProfileCard: React.FC<ProfileCardProps> = ({
  profile,
  isLoading
}) => {
  if (isLoading && !profile) {
    return <CardSkeleton />
  }

  const { character, name, profilePath } = profile
  return (
    <div className='max-w-sm min-w-[236px] w-full overflow-hidden shadow-md rounded-lg h-[350px] mx-auto p-2'>
      <div className='overflow-hidden rounded-xl relative flex flex-col transform hover:-translate-y-2 transition ease-in-out duration-500 shadow-lg hover:shadow-2xl movie-item text-white-primary h-[350px]'>
        <div className='absolute inset-0 z-10 transition duration-300 ease-in-out bg-gradient-to-t from-black-primary via-black-opacity to-transparent'></div>
        <div
          className='relative h-full group z-10 px-6 pt-10 pb-6 space-y-6 movie_info'
          data-lity=''
        >
          <div className='flex flex-col justify-end h-full'>
            <div className='text-center'>
              <p className='text-lg font-bold'>{name}</p>
              {character && (
                <div>
                  <span className='font-extralight'>Character: </span>
                  <span className='font-medium text-red-primary mt-1'>
                    {character}
                  </span>
                </div>
              )}
            </div>
          </div>
        </div>
        <img
          className='absolute inset-0 transform w-full -translate-y-4 object-cover h-full'
          src={getImage(profilePath!)}
          style={{ filter: 'grayscale(0)' }}
          alt='poster-image'
          data-testid='poster-image'
        />
      </div>
    </div>
  )
}
