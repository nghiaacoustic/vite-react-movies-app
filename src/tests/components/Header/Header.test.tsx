import { describe, it, expect } from 'vitest'
import { screen } from '@testing-library/react'
import { Route } from 'react-router-dom'
import { Header } from '@/components'
import { customRender } from '@/tests/mocks'

describe('Header component', () => {
  const renderComponent = () => {
    return customRender(<Route path='/' element={<Header />} />, ['/'])
  }

  it('should contain Home and Movies menu items', () => {
    renderComponent()

    const navHome = screen.getByText('Home')
    const navMovies = screen.getByText('Movies')

    expect(navHome).toBeInTheDocument()
    expect(navMovies).toBeInTheDocument()
  })

  it('should have correct class names for menu items', () => {
    renderComponent()

    const navHomeClass = screen.getByText('Home').getAttribute('class')
    const navMoviesClass = screen.getByText('Movies').getAttribute('class')

    expect(navHomeClass).toBe(
      'font-bold text-white-primary text-xl text-white-secondary hover:text-white-hover'
    )
    expect(navMoviesClass).toBe(
      'font-thin text-white-secondary text-xl text-white-secondary hover:text-white-hover'
    )
  })

  it('should match snapshot', () => {
    const { asFragment } = renderComponent()
    expect(asFragment()).toMatchSnapshot()
  })
})
