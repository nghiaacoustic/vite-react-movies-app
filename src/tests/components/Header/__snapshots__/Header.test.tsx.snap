// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > should match snapshot 1`] = `
<DocumentFragment>
  <header
    class="fixed flex-items-center justify-between py-4 px-8 w-full bg-gradient-to-r from-black-secondary via-transparent to-transparent transition duration-300 ease-in-out top-0 z-50"
  >
    <div
      class="flex-items-center"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        <img
          alt="Logo"
          class="h-10 mr-10 cursor-pointer"
          src="/src/assets/fpt-logo.png"
        />
      </a>
      <nav
        class="ml-auto"
      >
        <ul
          class="flex space-x-4"
        >
          <li>
            <a
              aria-current="page"
              class="font-bold text-white-primary text-xl text-white-secondary hover:text-white-hover"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="font-thin text-white-secondary text-xl text-white-secondary hover:text-white-hover"
              href="/movies/latest"
            >
              Movies
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>
</DocumentFragment>
`;
